specs/networking/p2p-interface.md

```diff
@@ -0,0 +1,743 @@
+- `proposer_slashing` - This topic is used solely for propagating proposer slashings to proposers on the network. Proposer slashings are sent in their entirety. Clients who receive a proposer slashing on this topic MUST validate the conditions within `process_proposer_slashing` before forwarding it across the network.
+- `attester_slashing` - This topic is used solely for propagating attester slashings to proposers on the network. Attester slashings are sent in their entirety. Clients who receive an attester slashing on this topic MUST validate the conditions within `process_attester_slashing` before forwarding it across the network.
+
+#### Interop
+
+Unaggregated attestations from all shards are sent to the `beacon_attestation` topic.
+
+#### Mainnet
+
+Shards are grouped into their own subnets (defined by a shard topic). The number of shard subnets is defined via `SHARD_SUBNET_COUNT` and the shard `shard_number % SHARD_SUBNET_COUNT` is assigned to the topic: `shard{shard_number % SHARD_SUBNET_COUNT}_beacon_attestation`. Unaggregated attestations are sent to the subnet topic. Aggregated attestations are sent to the `beacon_attestation` topic.
+
+### Messages
+
+Each gossipsub [message](https://github.com/libp2p/go-libp2p-pubsub/blob/master/pb/rpc.proto#L17-L24) has a maximum size of `GOSSIP_MAX_SIZE`.
+
+Clients MUST reject (fail validation) messages that are over this size limit. Likewise, clients MUST NOT emit or propagate messages larger than this limit.
+
+The payload is carried in the `data` field of a gossipsub message, and varies depending on the topic:
+
+
+| Topic                        | Message Type      |
+|------------------------------|-------------------|
+| beacon_block                 | BeaconBlock       |
+| beacon_attestation           | Attestation       |
+| shard{N}\_beacon_attestation | Attestation       |
+| voluntary_exit               | VoluntaryExit     |
+| proposer_slashing            | ProposerSlashing  |
+| attester_slashing            | AttesterSlashing  |
+
+Clients MUST reject (fail validation) messages containing an incorrect type, or invalid payload.
+
+When processing incoming gossip, clients MAY descore or disconnect peers who fail to observe these constraints.
+
+### Encodings
+
+Topics are post-fixed with an encoding. Encodings define how the payload of a gossipsub message is encoded.
+
+#### Interop
+
+- `ssz` - All objects are SSZ-encoded. Example: The beacon block topic string is: `/beacon_block/ssz` and the data field of a gossipsub message is an ssz-encoded `BeaconBlock`.
+
+#### Mainnet
+
+- `ssz_snappy` - All objects are ssz-encoded and then compressed with snappy. Example: The beacon attestation topic string is: `/beacon_attestation/ssz_snappy` and the data field of a gossipsub message is an `Attestation` that has been ssz-encoded then compressed with snappy.
+
+Implementations MUST use a single encoding. Changing an encoding will require coordination between participating implementations.
```

Comment:
Can't we handle encoding support in protocol negotiation? Seems like it would be a good idea to require clients to support ssz snappy, but allow them to support other encoding if they choose.

Reply:
The rationale for this was that segregating topics is similar to segregating nodes based on protocol id. Currently, nodes should not support two encodings. If they did, they would form two segregated networks and blocks propagating from one network to the other would be dependent on relay nodes (nodes that support both) and this can cause all kinds of headaches. 
For now, I think its best to avoid having relay/translation nodes and disjoint networks and just enforce everyone agrees on one compression. If the compression changes, we coordinate a hard fork.

Reply:
Yeah. Multiple encodings cause fragmentation, split brain, chimera; essentially diverging states. It may be a desirable property, but implementing it this way opens a can of worms, which can be addressed, but the cost/benefit isn’t clear to me at this point.

Reply:
We can still allow clients to publish multiple encodings as long as they at least publish the mandatory one - need a phrasing for that.

Reply:
Pubsub topics should be related to a type of content that is disseminating. In our case message encoding is not a part of the content itself, it rather a transport specific thing that matters between a pair of peers speaking to each other. Therefore, specifying encoding in topic name looks odd and puts additional complexity to the process of its update. It would require either multiple topic distribution from at least _some_ peers in the network or a hard fork coordination which could be quite a tricky thing. We might not want to have a deal with one or another.

Is there a rationale behind _not_ including encoding into a protocol?

Reply:
it's in the topic name because we're describing about the payload. 

broadcasts are generally not negotiated dynamically between two peers because that would require relaying/translation between different encodings and breaking gossip message identity (which determines if a message is retransmitted to neighbours or not). instead we're opting for a network-wide mandated encoding which will be upgraded only in a coordinated way (for example together with a hard fork). _insert radio station analogy_

there's a separate discussion to solve compression at the transport layer, but for the purposes of this spec, we're not there yet.

Reply:
> and breaking gossip message identity

Looks like I'm starting to get into this. We would have to have a deal with multiple identities for the same message in order to be able to gossip them across peers with different encodings. That is the point, right?

Reply:
well, different encodings on different topics means the message would have to be re-encoded by relays to reach the peers that listen only on the "other" encoding - this is the fragmentation that @raulk mentions above. in the process, you get differences in identity, so to know if a message has been published on the "other" encoding you have to reencode it first and check if you've seen it already. 

this is quite costly and non-trivial, so instead we're opting for mandating that all gossip happens on one topic with one encoding and upgrades are done "network-wide". 

also, gossip protocols typically remember what messages they've seen for a finite period of time - if you publish the same message again after that time has passed, it will be re-broadcast - adding a relay delay makes this scenario more likely (the identity problem, from a different angle).

one can imagine that in some complicated upgrade scenario, we might have peers publishing the same message on two topics/encodings, but the price here is pretty high in terms of overhead - both computational and networking, so we'd rather avoid that.

Reply:
> this is quite costly and non-trivial, so instead we're opting for mandating that all gossip happens on one topic with one encoding and upgrades are done "network-wide".

Would you mind to add a couple of sentences with this detail to the rationale section relating to topic encodings to prevent arising such questions from future readers?

Reply:
514ff5814f233836eba8b45d9bb5d94b97d7a60e